python/ray/data/datasource/parquet_datasource.py

```diff
@@ -257,12 +283,35 @@ def __init__(
                     "scheduling_strategy"
                 ] = DataContext.get_current().scheduling_strategy
 
-            self._metadata = (
-                meta_provider.prefetch_file_metadata(
-                    pq_ds.fragments, **prefetch_remote_args
-                )
-                or []
+            metadata_list = meta_provider.prefetch_file_metadata(
+                pq_ds.fragments, **prefetch_remote_args
             )
+            # Mapping of unique metadata objects ->
+            # set(fragment indices with that metadata)
+            self._unique_metadata = {}
+            # List of metadata objects corresponding to each fragment. References
+            # common metadata objects stored in self._unique_metadata in order to
+            # conserve memory by deduplicating common metadata objects across
+            # multiple fragments.
+            self._metadata = []
+
+            if metadata_list:
+                self._metadata = [None for _ in range(len(metadata_list))]
+                for fragment_idx, fragment_md in enumerate(metadata_list):
+                    # pyarrow.parquet.FileMetaData is not hashable, so we need to
+                    # serialize it to compare equality and store in dict.
+                    fragment_md = _SerializedFragmentMetadata(fragment_md)
+                    md_exists = False
+                    for unique_md, md_indices in self._unique_metadata.items():
```

Comment:
since we've already serialized the FileMetaData objects to binaries, we can simply put the binaries in a set to dedup them, no need to have this linear scan.